src/webapi/Controllers/AccountController.cs

```diff
@@ -0,0 +1,217 @@
+                {
+                    return new JsonResult(ModelState);
+                }
+
+                return Created($"/{Route}/{user.Id}", UserAccountResponse.From(user));
+            }
+
+            AddErrors(result);
+        }
+
+        return BadRequest(new { validationError = ModelState });
+
+
+        async Task<bool> AddToRole(ApplicationUser user)
+        {
+            var role = Roles.MentorTeacher;
+            if (AppSettings.Authentication.NewUsersAreAdministrators)
+            {
+                role = Roles.Administrator;
+            }
+
+            var roleResult = await _userManager.AddToRoleAsync(user, role);
+            if (!roleResult.Succeeded)
+            {
+                AddErrors(roleResult);
+                Response.StatusCode = StatusCodes.Status500InternalServerError;
+            }
+
+            return roleResult.Succeeded;
+        }
+    }
+
+
+    [HttpPut("{id}")]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<IActionResult> Put([FromRoute] string id, [FromBody] UserAccountUpdateRequest model)
+    {
+        if (ModelState.IsValid)
+        {
+            if (id != model.Id)
+            {
+                ModelState.AddModelError("id", "Id in URL does not match the Id in the message body.");
+                return BadRequest(ModelState);
+            }
+
+            var lookup = await _userManager.FindByIdAsync(id);
+            if (lookup is null || lookup.DeletedAt is not null)
+            {
+                return NotFound();
+            }
+
+            lookup.FirstName = model.FirstName;
+            lookup.LastName = model.LastName;
+            lookup.RequirePasswordChange = model.RequirePasswordChange;
+
+            if (model.EmailHasChanged(lookup))
+            {
+                lookup.UserName = lookup.Email = model.Email;
+                var updateEmailResult = await _userManager.UpdateAsync(lookup);
+                if (updateEmailResult.Succeeded)
+                {
+                    return NoContent();
+                }
+
+                AddErrors(updateEmailResult);
+            }
+            else
+            {
+                // The UpdateAsync function always tries to replace the email
+                // address. If you haven't changed the email address, it fails.
+                // Therefore we need a workaround for changing properties
+                // _other than_ email address.
+                var updateResult = await _identityRepository.Update(lookup);
+                if (!updateResult)
+                {
+                    // This means the user wasn't found, which shouldn't happen given that we have a lookup above.
+                    throw new InvalidOperationException("User update failed because user no longer exists.");
```

Comment:
`InvalidOpereationException` is the right exception to throw when something that shouldn't happen does happen.